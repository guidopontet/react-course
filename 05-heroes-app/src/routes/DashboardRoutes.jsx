import { Route, Routes } from "react-router-dom"

import { DcScreen } from "../components/dc/DcScreen"
import { HeroScreen } from "../components/hero/HeroScreen"
import { MarvelScreen } from "../components/marvel/MarvelScreen"
import { Navbar } from "../components/ui/Navbar"
import { SearchScreen } from "../components/search/SearchScreen"

export const DashboardRoutes = () => {
  return (
    <>
      <Navbar />

      <div className="mx-5 my-2">
        <Routes>
          <Route path="marvel" element={<MarvelScreen />} />
          <Route path="dc" element={<DcScreen />} />
          <Route path="search" element={<SearchScreen />} />
          <Route path="hero/:id" element={<HeroScreen />} />

          <Route path="/" element={<MarvelScreen />} />
        </Routes>
      </div>
    </>
  )
}
