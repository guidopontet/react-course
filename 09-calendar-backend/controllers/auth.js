const { response, request } = require('express');
const bcrypt = require('bcryptjs');

const User = require('../models/User');
const { generateJWT } = require('../helpers/jwt');

const createUser = async (req = request, res = response) => {
  const { email, password } = req.body;

  try {
    let user = await User.findOne({ email });

    if (user) {
      return res.status(400).json({
        ok: false,
        msg: 'User already exists',
      });
    }

    user = new User(req.body);

    const salt = bcrypt.genSaltSync();
    user.password = bcrypt.hashSync(password, salt);

    await user.save();

    const token = await generateJWT(user._id, user.name);

    res.status(201).json({
      ok: true,
      uid: user._id,
      name: user.name,
      token,
    })
  } catch (error) {
    res.status(500).json({
      ok: false,
      msg: 'Error',
      error,
    })
  }
}

const loginUser = async (req = request, res = response) => {
  const { email, password } = req.body;

  try {
    const user = await User.findOne({ email });

    if (!user) {
      return res.status(400).json({
        ok: false,
        msg: 'El usuario no existe',
      });
    }

    const validPassword = bcrypt.compareSync(password, user.password);

    if (!validPassword) {
      return res.status(400).json({
        ok: false,
        msg: 'El password no es correcto',
      });
    }

    const token = await generateJWT(user._id, user.name);

    res.json({
      ok: true,
      uid: user._id,
      name: user.name,
      token,
    });
  } catch (error) {
    res.status(500).json({
      ok: false,
      msg: 'Error',
      error,
    })
  }
}

const renewToken = async (req = request, res = response) => {
  const { uid, name } = req;

  const token = await generateJWT(uid, name);

  res.json({
    ok: true,
    token,
  })
}

module.exports = {
  createUser,
  loginUser,
  renewToken,
}
