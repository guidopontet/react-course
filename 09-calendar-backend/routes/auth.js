const { Router } = require('express');
const { check } = require('express-validator');

const { createUser, loginUser, renewToken } = require('../controllers/auth');
const { fieldsValidator } = require('../middlewares/fields-validator');
const { validateJWT } = require('../middlewares/validate-jwt');

const router = Router();

router.post(
  '/new',
  [
    check('name', 'El nombre es obligatorio').not().isEmpty(),
    check('email', 'El email es obligatorio').isEmail(),
    check('password', 'La contraseña debe ser de al menos 6 caracteres').isLength({ min: 6 }),
    fieldsValidator
  ],
  createUser
);

router.post(
  '/',
  [
    check('email', 'El email es obligatorio').isEmail(),
    check('password', 'La contraseña debe ser de al menos 6 caracteres').isLength({ min: 6 }),
    fieldsValidator,
  ],
  loginUser
);
router.get(
  '/renew',
  [ validateJWT ],
  renewToken
);

module.exports = router;
