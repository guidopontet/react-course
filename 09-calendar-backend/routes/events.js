const { Router } = require('express');
const { check } = require('express-validator');

const { fieldsValidator } = require('../middlewares/fields-validator');
const { getEvents, createEvent, updateEvent, deleteEvent } = require('../controllers/events');
const { isDate } = require('../helpers/isDate');
const { validateJWT } = require('../middlewares/validate-jwt');

const router = Router();

router.use(validateJWT); // Validate JWT on all routes below

router.get(
  '/',
  getEvents
);

router.post(
  '/',
  [
    check('title', 'El título es obligatorio').not().isEmpty(),
    check('start', 'La fecha de inicio es obligatoria').custom(isDate),
    check('end', 'La fecha de finalización es obligatoria').custom(isDate),
    fieldsValidator
  ],
  createEvent
);

router.put(
  '/:id',
  [
    check('title', 'El título es obligatorio').not().isEmpty(),
    check('start', 'La fecha de inicio es obligatoria').custom(isDate),
    check('end', 'La fecha de finalización es obligatoria').custom(isDate),
    fieldsValidator
  ],
  updateEvent
);

router.delete(
  '/:id',
  deleteEvent
);

module.exports = router;
