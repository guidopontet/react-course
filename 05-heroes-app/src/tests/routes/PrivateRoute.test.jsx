import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";

import { AuthContext } from "../../auth/authContext";
import { PrivateRoute } from "../../routes/PrivateRoute";

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Navigate: () => <span>Redireccionando</span>
}));

describe('PrivateRoute', () => {
  Storage.prototype.setItem = jest.fn();

  test('should render component if logged and save in localStorage', () => {
    const contextValue = {
      user: {
        logged: true,
        name: 'Guido'
      }
    }

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/']}>
          <PrivateRoute>
            <h1>Private component</h1>
          </PrivateRoute>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.text().trim()).toBe('Private component');
    expect(localStorage.setItem).toHaveBeenCalledWith('lastPath', '/');
  })

  test('should block component if not authenticated', () => {
    const contextValue = {
      user: {
        logged: false,
      }
    }

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/']}>
          <PrivateRoute>
            <h1>Private component</h1>
          </PrivateRoute>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wrapper.text().trim()).toBe('Redireccionando');
  })
})
