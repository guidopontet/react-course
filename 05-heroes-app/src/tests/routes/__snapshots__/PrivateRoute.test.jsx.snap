// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute should render component if logged and save in localStorage 1`] = `
<h1>
  Private component
</h1>
`;
